operators/src/pro/ml/xgboost.rs

```diff
@@ -0,0 +1,694 @@
+                global_geo_transform,
+                predicted_grid.into(),
+                props.clone(),
+            );
+
+        Ok(rt)
+    }
+}
+
+fn process_tile(
+    bands_of_tile: &Vec<f32>,
+    pool: &ThreadPool,
+    model_file: &[u8],
+    grid_shape: GridShape<[usize; 2]>,
+    n_bands: usize,
+    nan_val: f32,
+) -> Result<geoengine_datatypes::raster::Grid<GridShape<[usize; 2]>, f32>, XGBoostModuleError> {
+    pool.install(|| {
+        // to get one row of data means taking (n_pixels * n_bands) elements
+        let chunk_size = grid_shape.number_of_elements() * n_bands;
+
+        let res: Vec<_> = bands_of_tile
+            .par_chunks(chunk_size)
+            .map(|elem| {
+                // get xgboost style matrices
+                let xg_matrix = DMatrix::from_col_major_f32(
+                    elem,
+                    mem::size_of::<f32>() * n_bands,
+                    mem::size_of::<f32>(),
+                    grid_shape.number_of_elements(),
+                    n_bands,
+                    -1, // TODO: add this to settings.toml: # of threads for xgboost to use
+                    nan_val,
+                )
+                .context(self::error::CreateDMatrix)?;
+
+                let mut out_dim: u64 = 0;
+
+                let bst =
+                    Booster::load_buffer(model_file).context(self::error::LoadBoosterFromFile)?;
+
+                // measure time for prediction
+                let predictions = bst.predict_from_dmat(
+                    &xg_matrix,
+                    &[grid_shape.number_of_elements() as u64, n_bands as u64],
+                    &mut out_dim,
+                );
+
+                match predictions {
+                    Ok(_) => Ok(predictions),
+                    Err(xg_err) => Err(XGBoostModuleError::PredictionError { source: xg_err }),
+                }
+            })
+            .flatten_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let predictions_flat: Vec<f32> = res.into_iter().flatten().collect();
+
+        Grid2D::new(grid_shape, predictions_flat).context(self::error::DataTypes)
+    })
+}
+
+#[async_trait]
+impl<Q, P> QueryProcessor for XgboostProcessor<Q, P>
+where
+    Q: QueryProcessor<Output = RasterTile2D<P>, SpatialBounds = SpatialPartition2D>,
+    P: Pixel,
+{
+    type Output = RasterTile2D<PixelOut>;
+    type SpatialBounds = SpatialPartition2D;
+
+    async fn query<'a>(
+        &'a self,
+        query: RasterQueryRectangle,
+        ctx: &'a dyn QueryContext,
+    ) -> Result<BoxStream<'a, Result<Self::Output>>> {
+        let mut band_buffer = Vec::new();
+
+        for band in &self.sources {
+            let stream = band.query(query, ctx).await?;
+            band_buffer.push(stream);
+        }
+
+        let svz = StreamVectorZip::new(band_buffer);
+
+        let svz_vec: Vec<_> = svz.collect().await;
+        let stream_of_tiles_with_zipped_bands = stream::iter(svz_vec).map(Ok).boxed();
```

Comment:
why do you collect the stream into a vec and then create another stream out of it? Why don't you use svz directly?

Reply:
Good point.
https://github.com/geo-engine/geoengine/pull/639/commits/db3caa18b70d8dafeafe566d4b301fa9cef14786